Certificate Compression (RFC8879)

tmshort opened the pull request:
Add support for brotli compression
Add support for ~~draft-ietf-tls-certificate-compression[-05]~~ RFC8879
Add support for callback API, which allows the user to handle either compression expansion or pre-compressed server certificates.
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [X] documentation is added or updated
- [X] tests are added or updated

tmshort commented:
Certificate compression is enabled by default; so, it had to be removed from a number of the unit tests because it changes the number of extensions and the messages. One of the info callback tests checks to see that the compressed certificate message is indeed used.

tmshort commented:
I could not find tests for zlib upon which to base brotli compression.

tmshort commented:
brotli is the compression method supported by Chromium.

tmshort commented:
The AppVeyor builds were failing because of `make doc-nits`, doesn't help that there's no compression documentation at all!

tmshort commented:
This works with the stable release of Chrome (75).
Wireshark 3.0.0 recognizes the `compress_certificate` option.

slontis commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,729 @@
+/*
+ * Copyright 1998-2016 The OpenSSL Project Authors. All Rights Reserved.
```
Is this date a copy/paste?

tmshort replied:
Yes, this file was based on the c_zlib.c file in the same directory. But yeah, it should be updated.

tmshort replied:
Since it's based on c_zlib.c, I'm leaving the start date as 1998.

slontis commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,729 @@
+
+static int brotli_stateful_init(COMP_CTX *ctx)
+{
+    struct brotli_state *state = OPENSSL_zalloc(sizeof(*state));
```
why not just typedef this?

tmshort replied:
The equivalent isn't typedef'd in c_zlib.c; it's also not a public structure that is expected to be used everywhere.

slontis commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,729 @@
+    size_t in_avail = ilen;
+    size_t out_avail = olen;
+
+    if (state == NULL)
```
why do you check the state for NULL here but not in other methods above?

tmshort replied:
Appears that `brotli_stateful_finish()` is the outlier here; all other functions check for `state` not being NULL. This was based on the original c_zlib.c code, which also did not check its state variable in the finish function.

slontis commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,729 @@
+
+typedef struct {
+    struct { /* input structure */
+        size_t avail_in;
```
Should the common data sit outside encode/decode structs?

slontis commented on include/openssl/crypto.h:
```diff
@@ -399,7 +399,7 @@ int CRYPTO_memcmp(const void * in_a, const void * in_b, size_t len);
 # define OPENSSL_INIT_NO_ADD_ALL_KDFS        0x10000000L
 # define OPENSSL_INIT_ADD_ALL_KDFS           0x20000000L
-/* FREE: 0x40000000L */
+/* OPENSSL_INIT_BROTLI                       0x40000000L */
```
If this is really going to be declared this way you might want to put a comment above it explaining why.

tmshort replied:
`OPENSSL_INIT_ZLIB` is declared the same way. 

P.S. Not arguing, just explaining. If you want a comment here, then a comment for zlib would be appropriate, but then again, the intent of this PR was not to fix zlib's faults!

slontis replied:
I guess I dont understand why it needs a flag and then doesnt use it directly..

tmshort replied:
It's defined in cryptlib_int.h, as it's meant to only be used internally. But this is the master list of OPENSSL_INIT flags, so it needs to be "reserved" here.
But... (BIG **BUT**) it does appear that I missed the actual use of this flag... I did all the internals, but not the use of the flag in a manner similar to OPENSSL_INIT_ZLIB. I _was_ thinking I might reuse OPENSSL_INIT_ZLIB, and make it a generic OPENSSL_INIT_COMP flag, but I didn't get around to that. I will have to look to see if sharing an init flag between ZLIB and BROTLI makes sense, or if I should use the flag I defined!

slontis replied:
So its a reserved for internal use flag.. Even just saying that would be good.
Those flags are nearly full :). I tried one outside the 32 bit range and bad things happened on 32 bit machines..

tmshort replied:
Apparently, I do use it (easier to search in the workspace...)

mspncp replied:
@slontis the flag is used used here:

https://github.com/openssl/openssl/blob/fb84ea70a0cca8c6e430906500b00ebbe4463f2c/crypto/init.c#L657-L664

slontis commented on ssl/packet.c:
```diff
@@ -412,6 +412,12 @@ unsigned char *WPACKET_get_curr(WPACKET *pkt)
 
+void WPACKET_advance_write(WPACKET *pkt, size_t len)
+{
+    pkt->written += len;
```
Does this need to do a bounds check?

tmshort replied:
Maybe? I don't believe the WPACKET APIs are public, so it's only used internally. As used herein, the WPACKET buffer is grown to account. Perhaps an additional comment indicating use?

slontis replied:
I think one of the main goals of this API is to make sure no buffer overwrites happen. You are probably better to do the check so it doesnt bite someone in the future.

tmshort commented:
@kaduk what's the status on this draft, as in, do you think it will go another round, or will it be published as-is (technically) as an RFC?

tmshort commented on ssl/statem/statem_lib.c:
```diff
@@ -2471,12 +2471,11 @@ int tls13_construct_compressed_certificate(SSL *s, WPACKET *pkt, CERT_PKEY *cpk,
-    if (!WPACKET_put_bytes_u24(pkt, comp_len))
+    /* Put length then move write pointer past certificate */
+    if (!WPACKET_put_bytes_u24(pkt, comp_len)
+            || WPACKET_advance_write(pkt, comp_len))
```
This should be `!WPACKET_advance_write()`

kaduk commented:
Per https://www.rfc-editor.org/current_queue.php#draft-ietf-tls-certificate-compression the corresponding document is in the RFC Production Center's publication queue, but is currently waiting for draft-kucherawy-rfc8478bis to be published (currently in IETF Last Call).

tmshort commented:
re: draft-kucherawy-rfc8478bis I didn't put in zstd support.
Most of the recent changes to the ID for certificate compression are just reference updates; nothing to the protocol

tmshort commented:
Fully updated for RFC8879

tmshort commented:
The "GitHub CI / sanitizers" seems to be failing on test `30-test_evp_pkey_provided` which appears to be unrelated to this PR.

t8m commented:
I've rerun the jobs as we do not usually see a failure in 30-test_evp_pkey_provided. Just the encoder-decoder regularly fails.

tmshort commented:
With `config  --debug enable-asan enable-ubsan enable-rc5 enable-md2 enable-ec_nistp_64_gcc_128` (the same as the sanitize CI)
My local builds have shown:
```
SUMMARY: AddressSanitizer: 2020 byte(s) leaked in 35 allocation(s).
../../util/wrap.pl ../../test/evp_pkey_provided_test ../../test/recipes/30-test_evp_pkey_provided => 1
not ok 1 - running evp_pkey_provided_test
30-test_evp_pkey_provided.t ........ 1/? ---------------------------------------
#   Failed test 'running evp_pkey_provided_test'
#   at test/recipes/30-test_evp_pkey_provided.t line 17.
30-test_evp_pkey_provided.t ........ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
30-test_pbelu.t .................... ok

tmshort commented:
I see the above on the master branch as well, commit 0cc0164d193f6071a9d06b2116a410f8139a7e3

tmshort commented:
Just rebased again. @mspncp @slontis this should be ready to go. @mattcaswell sent out an email about getting PRs approved before beta, so I'm pushing for this one...

mattcaswell commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,759 @@
+
+static int brotli_stateful_compress_block(COMP_CTX *ctx, unsigned char *out,
+                                        unsigned int olen, unsigned char *in,
+                                        unsigned int ilen)
```
Indentation

tmshort replied:
Fixed, along with other places.

mattcaswell commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,759 @@
+#  endif
+
+    if (brotli_encode_dso == NULL)
+        brotli_encode_dso = DSO_load(NULL, LIBBROTLIENC, NULL, 0);
```
Yikes....race conditions here writing to a global. There probably should be some RUN_ONCE usage here.

tmshort replied:
This is based on the ZLIB code... similar problem...

tmshort replied:
I added RUN_ONCE to all the compression algorithms (including ZLIB) DSO support. It's done at the call of this function.

mattcaswell commented on crypto/comp/c_brotli.c:
```diff
@@ -0,0 +1,759 @@
+    if (ctx->decode.buf == NULL) {
+        ctx->decode.buf = OPENSSL_malloc(ctx->decode.bufsize);
+        if (ctx->decode.buf == NULL) {
+            COMPerr(COMP_F_BIO_BROTLI_READ, ERR_R_MALLOC_FAILURE);
```
We shouldn't be adding new `_F_` codes

FdaSilvaYY replied:
API usage has changed :
`        ERR_raise(ERR_LIB_COMP, ERR_R_MALLOC_FAILURE);`
There is still a wrapper to define `COMPerr`

tmshort replied:
Yes, the wrapper caught me here. I know I changed a log of things to be `ERR_raise()` as I dragged this PR forward.

tmshort replied:
> We shouldn't be adding new _F_ codes

No new _F_ codes were added due to the wrappers.

mattcaswell reviewed (changes requested):
I started to review this, but quickly realised its quite huge. I am concerned about adding such a large feature this late in the development cycle. For this reason I'm adding an OTC hold so that it can be discussed.

mattcaswell commented:
Also added it to the beta1 milestone to make sure it gets some OTC cycles.

tmshort commented:
> I started to review this, but quickly realised its quite huge. I am concerned about adding such a large feature this late in the development cycle. For this reason I'm adding an OTC hold so that it can be discussed.

This PR was opened in June 2019, ~21 months ago.

tmshort commented on include/openssl/ssl.h.in:
```diff
@@ -330,6 +330,9 @@ typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
 /* Allow initial connection to servers that don't support RI */
 # define SSL_OP_LEGACY_SERVER_CONNECT                    0x00000004U
 
+/* Disable certificate compression */
```
Thx. I had originally implemented this when it was a draft...

t-j-h commented on INSTALL.md:
```diff
@@ -388,8 +388,27 @@ for OpenSSL development.  It only works when using gcc or clang as the compiler.
+**On Unix**: this is the directory containing the brotli library.
+If not provided the system library path will be used.
+
+**On Windows:** no support.
```
Why no support on Windows?

tmshort replied:
I'm not a Windows developer. I would accept any help getting brotli supported on Windows.
Note: this is a brotli limitation, not a certificate compression limitation.

tmshort replied:
Windows support is now there; Windows CI has been updated to build/test with zstd and brotli.

t-j-h commented on INSTALL.md:
```diff
@@ -415,6 +434,28 @@ then this flag is optional and defaults to `ZLIB1` if not provided.
+[zstd-dynamic](#zstd-dynamic) option is not also used. If `zstd-dynamic` is used
+then this flag is optional and defaults to `LIBZSTD` if not provided.
+
+**On VMS:** support TBD.
```
We shouldn't have a PR with "TBD" in it like this.

kaduk replied:
Well, we shouldn't *merge* PRs with "TBD".  But we shouldn't expect generic contributors to know anything about VMS, either, and if the contributor is going to ask for help figuring out what or whether to do anything for VMS as part of the PR, does it matter if the string "TBD" is used to effectuate that or not?

tmshort replied:
This should really be "No support"; I'm not a VMS developer. I know that @levitte has been the driver for VMS support, I would accept any help from VMS support.

richsalz replied:
I would delete 457 and let interested parties add VMS support afterwards, if they care to do so.

mattcaswell replied:
> I would delete 457 and let interested parties add VMS support afterwards, if they care to do so.

This seems like a reasonable way to handle it

tmshort replied:
Deleted.

tmshort commented:
Rebased onto latest master - Fixed the comments.
Added CI for Unix/Windows to build/test with zstd and brotli (external and fuzz don't work, which is true on master as well).
Added RUN_ONCE to all the compression algorithm DSO support (including ZLIB).

tmshort commented:
Also updated to SSL_OP value to avoid a collision with KTLS.

mattcaswell commented:
OTC: This change is too significant for inclusion in 3.0. It can be considered for post 3.0.

FdaSilvaYY commented:
> OTC: This change is too significant for inclusion in 3.0. It can be considered for post 3.0.

At least, the "zlib loading by RUN_ONCE" should be submitted for 3.0, no ?
It will be a shorter PR to review and merge.

mattcaswell commented:
> At least, the "zlib loading by RUN_ONCE" should be submitted for 3.0, no ?

That would make sense

tmshort commented:
> > At least, the "zlib loading by RUN_ONCE" should be submitted for 3.0, no ?
> 
> That would make sense

PR #14940

xpahos commented on include/openssl/ssl.h.in:
```diff
@@ -329,6 +329,10 @@ typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
 # define SSL_OP_TLSEXT_PADDING                           0x00000010U
+
+/* Disable RFC8879 certificate compression */
+# define SSL_OP_NO_CERTIFICATE_COMPRESSION               0x00000020U
```
OpenSSL flags are very limited. Maybe it will be better to use an additional function to store enabled compression algorithms in SSL/SSL_CTX->ext?

tmshort replied:
The SSL_OP flags tie in with the configuration file nicely; that is why they are there. In addition, a number of flags become available in 3.0 due to deprecation of old features.

xpahos replied:
Maybe add something like **SSL_OP_CIPHER_SERVER_PREFERENCE** to force the client to use brotli in case of zlib first in the list but brotli or zstd on the second position?

tmshort replied:
An API to specify the the preference crypto order would probably work better. Your suggestion for promoting a single algorithm is very limited.

xpahos commented on ssl/statem/statem_lib.c:
```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
+        || !BUF_MEM_grow(*buf, expected_length)
+        || !PACKET_buf_init(tmppkt, (unsigned char*)(*buf)->data, expected_length)
+        || (comp = COMP_CTX_new(method)) == NULL
+        || COMP_expand_block(comp, (unsigned char*)(*buf)->data, expected_length,
```
> uncompressed_length:  The length of the Certificate message once it
>       is uncompressed.  If, after decompression, the specified length
>       does not match the actual length, the party receiving the invalid
>       message MUST abort the connection with the "bad_certificate"
>       alert.

Maybe **SSL_R_SSLV3_ALERT_BAD_CERTIFICATE** will be better here?

tmshort replied:
That might be an appropriate alert.

tmshort replied:
Based on a recent reading of RFC8879, this alert should be used here, and in a few other places.

xpahos commented on ssl/statem/statem_srvr.c:
```diff
@@ -3591,6 +3621,21 @@ MSG_PROCESS_RETURN tls_process_client_certificate(SSL *s, PACKET *pkt)
+    if (tls13_process_compressed_certificate(s, pkt, &tmppkt, &buf) != MSG_PROCESS_ERROR)
+        ret = tls_process_client_certificate(s, &tmppkt);
+
+    BUF_MEM_free(buf);
```
This is more of a niggle. Maybe it will be better to merge this function to prevent memory allocation inside other function and deallocation on this level. This code adds complexity to memory management.

tmshort replied:
Possibly valid point, I would have to go look at the the `buf` handling.

xpahos commented on ssl/statem/extensions.c:
```diff
@@ -1676,3 +1690,91 @@ static int init_post_handshake_auth(SSL *s, ossl_unused unsigned int context)
+        return 1;
+
+    if (s->ext.compress_certificate_rx != TLSEXT_comp_cert_none) {
+        SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_DUPLICATE_EXTENSION);
```
I'm not sure that an SSL connection could be used as a client and server at the same time. Does **SSL_R_DUPLICATE_EXTENSION** a proper error for this case?

tmshort replied:
This is the case where the peer sent two of the same extension. It is possible that a bad server or a bad client could send a duplicate extension. This code is reporting that there's a bug in the peer implementation. This has nothing to do with a connection being used simultaneously as server and client (which is not supported),

xpahos replied:
Sorry, you are right. Don't know why I was thinking that part of the code should be executed on the server side.

xpahos reviewed (commented):
Sorry for my English. Maybe it sounds rude. Just wanted to add some notes about the code.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -3616,6 +3661,18 @@ int tls_construct_server_certificate(SSL *s, WPACKET *pkt)
+#if !defined(OPENSSL_NO_COMP)
+int tls_construct_server_compressed_certificate(SSL *s, WPACKET *pkt)
+{
+    if (s->s3.tmp.cert == NULL) {
```
No, the state machine has to be able to create a compressed certificate message or a regular certificate message depending on the ClientHello. Just because the server may do a regular certificate message in one connection does not mean that it will be done in all connections. Both functions need to exist simultaneously.

tmshort commented on ssl/statem/statem_lib.c:
```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
+        goto err;
+    }
+
+    /* Use the |tmppkt| for the to-be-compressed data */
```
This code is not called when a regular server certificate is being generated. This is for compressed certificates which must compress the certificate; which is something that `tls_construct_server_certificate()` does not do.

xpahos replied:
Yes, this code called only on receiving the Compressed Certificate message. The first part of function duplicates **tls_construct_server_certificate** logic. Here you can generate a Certificate Message by calling **tls_construct_server_certificate** inside **tls13_construct_compressed_certificate** and than compress result of this function. You can just pass ```WPACKET tmppkt;```  to **tls_construct_server_certificate** and get the same result as the first 20 lines of your code.

tmshort replied:
You are proposing using the same buffer for both compression input and output, that is dangerous when the compression could possibly expand the output (which you don't know until after compression occurs); that is not safe.

It is safer to to get the data to be compressed in a separate buffer then compress into the output buffer (which is the buffer for the message).

xpahos replied:
https://github.com/xpahos/openssl/blob/dcdc35357fbb4c81066b8524bc1192f41a5e2e6b/ssl/statem/statem_srvr.c#L3650

I think it will be better to show in the code what I want to say :)

tmshort commented on ssl/statem/statem_lib.c:
```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
+
+    /* continue with the real |pkt| */
+    if (!WPACKET_put_bytes_u16(pkt, s->ext.compress_certificate_rx)
+            || !WPACKET_get_total_written(&tmppkt, &length)
```
This code is not called when a regular server certificate is being generated. This is for compressed certificates which must compress the certificate; which is something that `tls_construct_server_certificate()` does not do. `tls_construct_server_certificate()` and `tls_construct_compressed_server_certificate()` are *not* called on the same connection.

tmshort replied:
Again, this is the case of a compressed certificate where `tls_constrict_server_certificate()` is not being called. Two separate messages, who separate construction functions.

tmshort commented on ssl/statem/statem_lib.c:
```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
+    switch (s->ext.compress_certificate_rx) {
+# ifdef ZLIB
+    case TLSEXT_comp_cert_zlib:
+        max_length = length + 11 + 5 * (length >> 14);
```
A valid point. But approximate size is "not good enough". The COMP API does not provide a mechanism to calculate the approximate size and not all algorithms support that.

That being said, the handshake is already at a spot where we've determined that the certificate is to be compressed, there's no good way to back out (the state machine really doesn't permit changing the message type while constructing a message).

CPU time would be wasted in determining the size of the compression first, then doing the actual compression. In effect, you would be doubling the CPU time by calculating the size first.

xpahos replied:
Maybe send an Alert message and drop connection in case of compressed size bigger than uncompressed. It's strange behaviour when you expecting a decreasing size of the message but got a message twice bigger than uncompressed. It will be really difficult to catch this problem without alerting from the server-side.

tmshort replied:
I consider that bad behavior. If it can work, it should work. Dropping the connection is a poor substitute for properly handling compression expansion. I would prefer some other diagnostic that indicates that compression expanded.

xpahos replied:
Maybe store somewhere information about the size of compressed, uncompressed and compression algorithm id? Somewhere in SSL structure. With this functionality, it could be possible to track size and add alerts on the server-side. Furthermore, it will be possible to graph the distribution of users with and without compression algorithms.

tmshort replied:
That seems a bit out of scope for the OpenSSL library, beyond simple per-connection stats (but, OpenSSL is not known for it's plethora of connection stats).

xpahos replied:
Unfortunately without any information about compressed certificate size, there is no way to detect the problem. BoringSSL uses callbacks for compression algorithms. This could be used as external cache storage and statistics about clients.

tmshort replied:
A callback API was added

tmshort commented on ssl/ssl_local.h:
```diff
@@ -1664,6 +1665,14 @@ struct ssl_st {
+         * If the peer supports certificate compression, this is what
+         * we will use: TLXEXT_comp_cert_XXX
+         */
+        int compress_certificate_rx;
```
True, but it is usually recommended to use `int` for simple storage in a structure, as it is generally aligned and faster to access regardless of architecture. These structures are not PACKED.
Consider *all* the other variables that are `int`s, but only store true/false or a few (enumerated) values.

tmshort commented:
Hoping this can be reviewed for 3.1? @mattcaswell @slontis @mspncp

FdaSilvaYY reviewed (commented):
few nits ;)

tmshort commented:
> few nits ;)

And English _is_ my native language... sigh... Thanks.

tmshort commented on apps/enc.c:
```diff
@@ -405,6 +430,36 @@ int enc_main(int argc, char **argv)
+        if (enc)
+            wbio = BIO_push(bzstd, wbio);
+        else
+            rbio = BIO_push(bzstd, rbio);
```
Yes. Not sure if there's a concern here or...?

slontis replied:
It seems like a strange thing to allow. I would not expect it to allow more that one of these at a time..

tmshort commented:
Zstd on windows seems to have intermittent problems downloading/installing; that's before the compile, so not directly related to a code change.

tmshort commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -162,6 +162,19 @@ This is a synonym for the B<-groups> command.
+
+Enables support for SSL/TLS compression, same as clearing
+B<SSL_OP_NO_CERTIFICATE_COMPRESSION>.
+This command was introduced in OpenSSL 3.0.0.
```
I _just_ went through the documentation to update for 3.1, but missed this!

tmshort commented:
github is hanging onto the "@mattcaswell request changes" status pretty hard.

tmshort commented:
@mattcaswell @paulidale @t8m How does one get this out of the `hold: substantial feature` state?

paulidale commented:
There will be a blog post about this.  These things can, unfortunately, take a while.

slontis commented on INSTALL.md:
```diff
@@ -19,7 +19,7 @@ Table of Contents
    - [Directories](#directories)
    - [Compiler Warnings](#compiler-warnings)
-   - [ZLib Flags](#zlib-flags)
+   - [Compression Algorithm Flags](#comp-alg-flags)
```
The #comp-alg-flags link does not map to the section name..

slontis commented on Configurations/00-base-templates.conf:
```diff
@@ -69,11 +77,22 @@ my %targets=(
         CC              => "cc",
         lflags          =>
-            sub { $withargs{zlib_lib} ? "-L".$withargs{zlib_lib} : () },
+            sub { my @libs = ();
```
Maybe use the same indentation as the above include block.. It is currently all over the place (same applies to ex_libs)

slontis commented on doc/man1/openssl-s_client.pod.in:
```diff
@@ -589,6 +591,17 @@ For more information on shutting down a connection, see L<SSL_shutdown(3)>.
 
+=item B<-cert_comp>
+
+Enables support for SSL/TLS certificate compression, same as clearing
```
Maybe TLS 1.3 instead? (Same comment applies to next section)

tmshort replied:
Fixed

slontis commented on doc/man3/SSL_COMP_add_compression_method.pod:
```diff
@@ -60,7 +60,21 @@ when a matching identifier is found. There is no way to restrict the list
 of compression methods supported on a per connection basis.
 
 If enabled during compilation, the OpenSSL library will have the
```
This might need to be a bit clearer now that there is multiple options.

slontis commented on doc/man1/openssl-s_server.pod.in:
```diff
@@ -599,6 +601,17 @@ further information).
 
+=item B<-cert_comp>
+
+Enables support for SSL/TLS certificate compression, same as clearing
```
TLS 1.3 (same applies to section below).

tmshort replied:
Fixed

slontis commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -162,6 +162,19 @@ This is a synonym for the B<-groups> command.
 
+=item B<-cert_comp>
+
+Enables support for SSL/TLS compression, same as clearing
```
TLS 1.3 (same applies to section below).

tmshort replied:
Fixed

slontis commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -533,6 +546,10 @@ B<KTLS>: Enables kernel TLS if support has been compiled in, and it is supported
 by the negotiated ciphersuites and extensions. Equivalent to
 B<SSL_OP_ENABLE_KTLS>.
 
+B<CertificateCompression>: use the certificate compression extension, enabled by
```
Will need to be decided if this should be on by default.

tmshort replied:
It depends if any compression algorithms are available, as well.

slontis commented on util/libcrypto.num:
```diff
@@ -5425,3 +5425,9 @@ ASN1_item_d2i_ex                        5552	3_0_0	EXIST::FUNCTION:
 ASN1_TIME_print_ex                      5553	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_get0_provider                  5554	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_CTX_get0_provider              5555	3_0_0	EXIST::FUNCTION:
+COMP_brotli                             ?	3_0_0	EXIST::FUNCTION:COMP
```
Should these be EXIST::FUNCTION:BROTLI,COMP

mattcaswell commented:
> @mattcaswell @paulidale @t8m How does one get this out of the hold: substantial feature state?

The point of having this hold is to identify significant features that are stuck because they are big and complex. The idea is that the OTC will first review them to confirm whether or not they are a good idea from a technical perspective before being passed to OMC for an assessment of priority (i.e. if and when they should go in). It's on the draft agenda for the next OTC meeting on Tuesday.

slontis commented:
This will be required for QUIC

tmshort commented:
check_docs appears to be complaining about an area of `CHANGES.md` untouched by this PR.

t8m commented:
> check_docs appears to be complaining about an area of `CHANGES.md` untouched by this PR.

Fix for that in #17274

tmshort commented:
> > check_docs appears to be complaining about an area of `CHANGES.md` untouched by this PR.
> 
> Fix for that in #17274

Thanks, I wasn't worried about it... but thought I'd note "it wasn't me!" :)

tmshort commented:
> The point of having this hold is to identify significant features that are stuck because they are big and complex. The idea is that the OTC will first review them to confirm whether or not they are a good idea from a technical perspective before being passed to OMC for an assessment of priority (i.e. if and when they should go in). It's on the draft agenda for the next OTC meeting on Tuesday.

Any updates from OTC @mattcaswell ?

paulidale commented:
The OTC didn't get to this (and others).  It's on the agenda for the next call (later today for me).

mattcaswell commented:
> The OTC didn't get to this (and others). It's on the agenda for the next call (later today for me).

Actually its already been passed to OMC

FdaSilvaYY commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -162,6 +162,19 @@ This is a synonym for the B<-groups> command.
+
+Enables support for TLSv1.3 certificate compression, same as clearing
+B<SSL_OP_NO_CERTIFICATE_COMPRESSION>.
+This command was introduced in OpenSSL 3.1.0.
```
Nit : This feature or flag was introduced... ?

FdaSilvaYY commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -460,7 +473,7 @@ B<SessionTicket>: session ticket support, enabled by default. Inverse of
 B<SSL_OP_NO_TICKET>.
 
-B<Compression>: SSL/TLS compression support, enabled by default. Inverse
+B<Compression>: SSL/TLS compression support, disabled by default. Inverse
```
If i remember well, TLS Compression is disabled by default, for security reasons.
So this is a documentation fix that should be backported ?

tmshort replied:
Yup. Since I was in here dealing with (other) compression, I fixed it. But this would be appropriate for backporting.

mattcaswell commented:
The OMC have voted to remove the current holds ("need OMC decision" and "substantial feature") from this PR. It may now proceed normally as per the standard review process.

paulidale commented:
Ping @openssl/committers for reviews.  I do not feel qualified to do anything better than a too superficial review of this.

tmshort commented:
Closing in favor of #18186
